File: GUI/WpfMidiOrganPedals/Devices/GeneralStatusMessage.cs
namespace WpfMidiOrganPedals.Devices
{
    internal class GeneralStatusMessage : Message
    {
        public const byte Id = 0x01;

        private uint pressedPedals;
        private uint playedNotes;
        private ushort numberOfToggledPedals;
        private ushort numberOfToggledNotes;
        private ushort numberOfDiscardedBytes;

        public GeneralStatusMessage(byte[] rawData)
        {
            var extractor = new RawMessageExtractor(rawData);
            extractor.Get(ref pressedPedals);
            extractor.Get(ref playedNotes);
            extractor.Get(ref numberOfToggledPedals);
            extractor.Get(ref numberOfToggledNotes);
            extractor.Get(ref numberOfDiscardedBytes);
            extractor.Finish();
        }

        public GeneralStatusMessage()
        {
            pressedPedals = 0b01010101010101010101010101010101;
            playedNotes = 0b10101010101010101010101010101010;
            numberOfToggledPedals = 33;
            numberOfToggledNotes = 34;
            numberOfDiscardedBytes = 123;
        }

        public uint PressedPedals => pressedPedals;
        public uint PlayedNotes => playedNotes;
        public ushort NumberOfToggledPedals => numberOfToggledPedals;
        public ushort NumberOfToggledNotes => numberOfToggledNotes;
        public ushort NumberOfDiscardedBytes => numberOfDiscardedBytes;

        public override RawMessage Pack()
        {
            var rawMessageBuilder = new RawMessageBuilder();
            rawMessageBuilder.Add(pressedPedals);
            rawMessageBuilder.Add(playedNotes);
            rawMessageBuilder.Add(numberOfToggledPedals);
            rawMessageBuilder.Add(numberOfToggledNotes);
            rawMessageBuilder.Add(numberOfDiscardedBytes);
            var rawData = rawMessageBuilder.GetData();
            var message = new RawMessage(Id, rawData);
            return message;
        }
    }
}

File: Firmware/GeneralStatusMessage.cpp
#include "GeneralStatusMessage.h"
#include "RawMessageBuilder.h"


GeneralStatusMessage::GeneralStatusMessage(
  bool configurationOk,
  uint32_t pressedPedals,
  uint32_t playedNotes,
  uint16_t numberOfToggledPedals,
  uint16_t numberOfToggledNotes,
  uint16_t numberOfReceivedBytes,
  uint16_t numberOfDiscardedBytes)
{
  this->configurationOk = configurationOk;
  this->pressedPedals = pressedPedals;
  this->playedNotes = playedNotes;
  this->numberOfToggledPedals = numberOfToggledPedals;
  this->numberOfToggledNotes = numberOfToggledNotes;
  this->numberOfReceivedBytes = numberOfReceivedBytes;
  this->numberOfDiscardedBytes = numberOfDiscardedBytes;
}


void GeneralStatusMessage::Pack(RawMessage &result)
{
  RawMessageBuilder packer;
  packer.Add(configurationOk);
  packer.Add(pressedPedals);
  packer.Add(playedNotes);
  packer.Add(numberOfToggledPedals);
  packer.Add(numberOfToggledNotes);
  packer.Add(numberOfReceivedBytes);
  packer.Add(numberOfDiscardedBytes);
  packer.CopyTo(Id, result);
}

